Added scroll bar to ListBox2D

karandeepSJ opened the pull request:
This PR adds a scroll bar to the `ListBox2D` in viz.ui. The scroll bar can be moved at different speeds with the mouse. It also moves up or down when the scrolling occurs through the mouse wheel.
![scroll](https://user-images.githubusercontent.com/24621165/41396599-19ab4cde-6fcf-11e8-9e84-7668a946f87b.gif)

pep8speaks commented:
Hello @karandeepSJ, Thank you for submitting the Pull Request !

 - In the file [`dipy/viz/ui.py`](https://github.com/nipy/dipy/blob/5fd9218b637263c8c4706a714c274be84bb77382/dipy/viz/ui.py), following are the PEP8 issues :

> [Line 2253:81](https://github.com/nipy/dipy/blob/5fd9218b637263c8c4706a714c274be84bb77382/dipy/viz/ui.py#L2253): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 80 characters)
> [Line 2267:81](https://github.com/nipy/dipy/blob/5fd9218b637263c8c4706a714c274be84bb77382/dipy/viz/ui.py#L2267): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)
> [Line 2275:81](https://github.com/nipy/dipy/blob/5fd9218b637263c8c4706a714c274be84bb77382/dipy/viz/ui.py#L2275): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)

Do see the [DIPY coding Style guideline](https://github.com/nipy/dipy/blob/master/doc/devel/coding_style_guideline.rst)

skoudoro commented:
Hi @karandeepSJ, Thanks for this cool features!  Can you check [this error](https://travis-ci.org/nipy/dipy/jobs/392114901#L4951) on travis?

skoudoro commented on dipy/viz/tests/test_ui.py:
```diff
@@ -528,41 +528,8 @@ def _on_change():
-    # Check if the right values were selected.
-    expected = [[1], [2], [2], [21], [1], [42]]
-    assert len(selected_values) == len(expected)
-    assert_arrays_equal(selected_values, expected)
```
Why are you removing this test?

Garyfallidis commented:
Same here @karandeepSJ . Needs rebasing.

Garyfallidis commented:
@karandeepSJ please resolve the conflicts of this PR.

dmreagan commented:
This looks pretty nice, but I've observed some small quirks in the scrollbar behavior. I found that if you drag the bar downward in small motions, it scrolls up instead of down. The reverse is also true. I could only trigger this behavior when the scrollbar is large.

![scrollbar_bug](https://user-images.githubusercontent.com/6951528/43288843-21236642-90f7-11e8-90b3-6bce419d38c7.gif)

dmreagan commented:
The bug is fixed, so this PR looks good to me. How difficult would it be to change the color of the bar when it is grabbed like we do with sliders? It could wait for another PR. Any thoughts here, @ranveeraggarwal and @MarcCote ?

MarcCote commented on dipy/viz/tests/test_ui.py:
```diff
@@ -539,13 +539,38 @@ def test_ui_range_slider(interactive=False):
 @npt.dec.skipif(not have_vtk or skip_it)
 @xvfb_it
-def test_ui_listbox_2d(recording=False):
+def test_ui_listbox_2d(interactive=False):
```
Like we discussed in PR #1559 , we may want to have a `mode` parameter that can either accep `None, `interactive` or `recording`.

MarcCote commented on dipy/viz/ui.py:
```diff
@@ -3009,41 +3014,57 @@ def _setup(self):
-        pos = (pos[0], self.up_button.size[1] // 2 + margin)
-        self.panel.add_element(self.down_button, pos, anchor="center")
+        # arrow_up = read_viz_icons(fname="arrow-up.png")
+        # self.up_button = Button2D([("up", arrow_up)])
```
Oh, let's remove those commented lines.

MarcCote commented on dipy/viz/ui.py:
```diff
@@ -3009,41 +3014,57 @@ def _setup(self):
         # Add default events listener for this UI component.
-        self.up_button.on_left_mouse_button_pressed = self.up_button_callback
-        self.down_button.on_left_mouse_button_pressed = self.down_button_callback
+        # self.up_button.on_left_mouse_button_pressed = self.up_button_callback
```
Remove commented lines.

karandeepSJ commented:
@dmreagan  This is how it looks like when the color of the scroll bar changes on grabbing it. Currently, I have only made it darker.

![scroll_color](https://user-images.githubusercontent.com/24621165/43962890-776b67ee-9cd6-11e8-9ed6-a9279a250ea3.gif)

karandeepSJ commented:
@dmreagan  @skoudoro  @Garyfallidis There were some connection problem during the travis build of this PR and of #1559.

dmreagan commented:
@karandeepSJ please rebase

karandeepSJ commented:
Rebased. To adjust the scroll bar size and step length on changing a directory, I added a function `update_scrollbar` to `ListBox2D'. I also added the `update_element` function to `Panel2D` to update the position of the scroll bar in the panel

dmreagan commented:
@MarcCote please take a look at @karandeepSJ's latest update. It looks like `update_element` is repeating a lot of code from `add_element`. Is there a better way to do this?

MarcCote commented:
@dmreagan @karandeepSJ we could refactor the `update_element` code like this
```python

    
    def remove_element(self, element):
        idx = self._elements.index(element)
        del self._elements[idx]
        del self.element_offsets[idx]

    def update_element(self, element, coords, anchor="position"):
        """ Updates the position of a UI component in the panel.

        Parameters
        ----------
        element : UI
            The UI item to be updated.
        coords : (float, float) or (int, int)
            New coordinates.
            If float, normalized coordinates are assumed and they must be
            between [0,1].
            If int, pixels coordinates are assumed and it must fit within the
            panel's size.
        """
        self.remove_element(element)
        self.add_element(element, coords, anchor)
```

skoudoro commented:
Hi @karandeepSJ, Any chance to have the improvement above proposed by @MarcCote  before merging this PR?

karandeepSJ commented:
Yes, I am working on it

karandeepSJ commented:
Done. Also, there was an error being thrown when I tried to open a directory without read permissions from the filemenu. I got rid of the error. Now the user just stays in the same directory

skoudoro commented:
Great! It is ready to be merged. Can you do it @dmreagan? 

Thank you @karandeepSJ!
